File: code/c++/search1.cc
#include <xapian.h>

#include <cstdio>
#include <iomanip>
#include <iostream>
#include <string>

using namespace std;

static string
get_field(const string & data, size_t field)
{
    size_t start = 0;
    while (true) {
	size_t end = data.find('\n', start);
	if (field == 0)
	    return string(data, start, end - start);
	start = end;
	if (start != string::npos) ++start;
	--field;
    }
}

// Start of example code.
static void
search(const string & dbpath, const string & querystring,
       Xapian::doccount offset = 0, Xapian::doccount pagesize = 10)
{
    // offset - defines starting point within result set.
    // pagesize - defines number of records to retrieve.

    // Open the database we're going to search.
    Xapian::Database db(dbpath);

    // Set up a QueryParser with a stemmer and suitable prefixes.
    Xapian::QueryParser queryparser;
    queryparser.set_stemmer(Xapian::Stem("en"));
    queryparser.set_stemming_strategy(queryparser.STEM_SOME);
    queryparser.add_prefix("title", "S");
    queryparser.add_prefix("description", "XD");

    // And parse the query.
    Xapian::Query query = queryparser.parse_query(querystring);

    // Use an Enquire object on the database to run the query.
    Xapian::Enquire enquire(db);
    enquire.set_query(query);

    // And print out something about each match.
    Xapian::MSet mset = enquire.get_mset(offset, pagesize);

    clog << "'" << querystring << "'[" << offset << ":" << offset + pagesize
	 << "] =";
    for (Xapian::MSetIterator m = mset.begin(); m != mset.end(); ++m) {
	Xapian::docid did = *m;
	cout << m.get_rank() + 1 << ": #" << setfill('0') << setw(3) << did
	     << ' ';

	const string & data = m.get_document().get_data();
	cout << get_field(data, 1) << endl;
	// Log the document id.
	clog << ' ' << did;
    }
    clog << endl;
}
// End of example code.

int main(int argc, char** argv) {
    if (argc < 3) {
	cerr << "Syntax: " << argv[0] << " DBPATH QUERYTERM[...]" << endl;
	return 1;
    }
    const char * dbpath = argv[1];

    // Join the rest of the arguments with spaces to make the query string.
    string querystring;
    for (argv += 2; *argv; ++argv) {
	if (!querystring.empty()) querystring += ' ';
	querystring += *argv;
    }

    search(dbpath, querystring);
}

File: code/java/search1.java
package code.java;

import org.xapian.Database;
import org.xapian.Document;
import org.xapian.Enquire;
import org.xapian.MSet;
import org.xapian.MSetIterator;
import org.xapian.Query;
import org.xapian.QueryParser;
import org.xapian.Stem;

public class search1 {

    // Command line args - dbpath querystring
    public static void main(String[] args) {
        if (args.length < 2) {
            System.out.println("Insufficient number of arguments (should be dbpath querystring)");
            return;
        }
        search(args[0], args[1]);
    }

    public static void search(String dbpath, String queryString) {
        search(dbpath, queryString, 0, 10);
    }

    // Start of example code.
    public static void search(String dbpath, String queryString, int offset, int pagesize) {
        // offset - defines starting point within result set
        // pagesize - defines number of records to retrieve

        // Open the databse we're going to search.
        Database db = new Database(dbpath);

        // Set up a QueryParser with a stemmer and suitable prefixes
        QueryParser queryParser = new QueryParser();
        queryParser.setStemmer(new Stem("en"));
        queryParser.setStemmingStrategy(QueryParser.stem_strategy.STEM_SOME);
        // Start of prefix configuration.
        queryParser.addPrefix("title", "S");
        queryParser.addPrefix("description", "XD");
        // End of prefix configuration.

        // And parse the query
        Query query = queryParser.parseQuery(queryString);

        // Use an Enquire object on the database to run the query
        Enquire enquire = new Enquire(db);
        enquire.setQuery(query);

        // And print out something about each match
        MSet mset = enquire.getMSet(offset, pagesize);
        MSetIterator msetIterator = mset.begin();

        while (msetIterator.hasNext())
        {
            long rank = msetIterator.getRank();
            long docID = msetIterator.getDocId();
            Document doc = db.getDocument(docID);

            System.out.printf("%i: #%3.3i %s%n", rank+1, docID, doc.getValue(0))
            msetIterator.next();
        }

        System.out.printf("'%s'[%i:%i] = ", queryString, offset, offset+pagesize);
        msetIterator = mset.begin();
        while (msetIterator.hasNext())
        {
            System.out.printf("%i", msetIterator.getDocId());
            msetIterator.next();
            if (msetIterator.hasNext()) {
                System.out.print(" ");
            }
        }
        System.out.print("\n");
    }
    // End of example code.
}
